sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/combiner/SequenceRangeAccumulator.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.ordered.combiner;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.extensions.ordered.ContiguousSequenceRange;
+import org.apache.commons.lang3.tuple.Pair;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.UnknownKeyFor;
+import org.joda.time.Instant;
+
+/** Default accumulator used to combine sequence ranges. */
+public class SequenceRangeAccumulator {
+
+  private static Instant max(Instant a, Instant b) {
+    return a.isAfter(b) ? a : b;
+  }
+
+  private final TreeMap<Long, Pair<Long, Instant>> data = new TreeMap<>();
```

Comment:
I'm having a hard time following what this does, could we add a comment here?

I'm understanding it basically as holding a bunch of disjoint sequences: `<lower end of range, <upper end of range, timestamp?>>`, but I don't really follow what the timestamp is. Maybe ranges would be a better name as well?

Reply:
I think the timestamp is the highest timestamp of any event we've seen, but I don't really understand why we need it

Reply:
This max timestamp is critical for correctness - it will be the time to which the event time domain based timers will be set in the processing ParDos. Beam runners will guarantee that all events up to that time are received by the DoFn, and it means that we have a complete set of messages for a given key for a given range.

Reply:
Got it - that makes sense, but is not intuitive (or at least wasn't to me) - could we add a comment here (and maybe consider a more descriptive variable name)

Reply:
Added the comments. I kept the variable names the same; let me know please if the comments are not sufficent.

Reply:
LGTM, thank you